torch/distributed/nn/functional.py

```diff
@@ -207,6 +232,20 @@ def backward(ctx, grad_output):
         return (None, None, None) + (_Broadcast.apply(ctx.src, ctx.group, grad_output),)
 
 
+class _Reduce_Scatter(Function):
+    @staticmethod
+    def forward(ctx, op, group, tensor, *input_tensor_list):
+        ctx.group = group
+        dist.reduce_scatter(tensor, list(input_tensor_list), op=op, group=group)
+        return tensor
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        result = _AllGather.apply(ctx.group, grad_output.contiguous())
```

Comment:
Do we need to make this contiguous?

Reply:
Unfortunately, yes. So in the unit test which I write in this PR, the test passes. But for sharded linear test cases, it throws this exception:

> RuntimeError: Tensors must be non-overlapping and dense.

So I have to add this. And I don't know whether I am missing anything here.

Reply:
Ah I see, do we know where this non-contiguous tensor comes from in sharded linear? We have mostly focussed on enabling functionality currently, but when we need to make the sharded linear implementation performant, we will have to look into things like these to avoid additional contiguous calls.

Also, if this is coming from sharded linear I'm wondering if it can be fixed there and in the core collective calls we don't have any `.contiguous()` calls?